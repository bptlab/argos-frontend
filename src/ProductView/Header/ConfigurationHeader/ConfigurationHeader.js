import React, { Component } from 'react';

class ConfigurationHeader extends Component {
    constructor(props) {
        super(props);

        this.buildCodingPlugAndSoftwareVersionToIdMapping();
        const initialCodingPlugVersion = Object.keys(this.cpSwToIdMapping)[0];
        this.state = {selectedCodingPlug: initialCodingPlugVersion,
                        selectedSoftwareVersion: Object.keys(this.cpSwToIdMapping[initialCodingPlugVersion])[0],
                        showAll: true};

        this.handleChangeSelectedCodingPlug = this.handleChangeSelectedCodingPlug.bind(this);
        this.handleChangeSelectedSoftwareVersion = this.handleChangeSelectedSoftwareVersion.bind(this);
        this.toggleShowAll = this.toggleShowAll.bind(this);
    }

    buildCodingPlugAndSoftwareVersionToIdMapping() {
        this.cpSwToIdMapping = {};
        for (const configuration of this.props.configurations) {
            if(!this.cpSwToIdMapping[configuration.codingPlugId]) {
                this.cpSwToIdMapping[configuration.codingPlugId] = {};
            }
            for (const softwareVersion of configuration.codingPlugSoftwareVersions) {
                this.cpSwToIdMapping[configuration.codingPlugId][softwareVersion] = configuration.id;
            }
        }
    }
    selectedSoftwareVersionChanged(event) {
        this.setState({selectedSoftwareVersion: event.target.value});
    }

    onChangeProductConfiguration(newCodingPlug, newSoftwareVersion) {
        const selectedConfigurationId = this.cpSwToIdMapping[newCodingPlug][newSoftwareVersion];
        this.props.onChangeProductConfiguration(selectedConfigurationId);
    }

    handleChangeSelectedCodingPlug(event) {
        const newCodingPlug = event.target.value;
        const newSoftwareVersion = Object.keys(this.cpSwToIdMapping[newCodingPlug])[0];
        this.setState({selectedCodingPlug: newCodingPlug,
            selectedSoftwareVersion: newSoftwareVersion});
        this.onChangeProductConfiguration(newCodingPlug, newSoftwareVersion);
    }

    handleChangeSelectedSoftwareVersion(event) {
        this.setState({selectedSoftwareVersion: event.target.value});
        this.onChangeProductConfiguration(this.state.selectedCodingPlug, event.target.value);
    }

    toggleShowAll(event) {
        this.setState({
            showAll: event.target.checked
        });
        if(event.target.checked) {
            this.props.onChangeProductConfiguration(null);
        }
        else {
            this.onChangeProductConfiguration(this.state.selectedCodingPlug, this.state.selectedSoftwareVersion);
        }
    }

    render() {
        return (
            <div className="row col-12 d-flex justify-content-center">
                <div className="col-4 d-flex justify-content-around">
                      <div className="form-group">
                          <input
                              type="checkbox"
                              id="showAllCheckbox"
                              checked={this.state.showAll}
                              onChange={this.toggleShowAll}/>
                          <label htmlFor="softwareVersionSelection">show all configurations</label>
                      </div>
                </div>
                <div className="col-3 d-flex">
                    <div className="form-group">
                        <label htmlFor="codingPlugSelection" hidden={this.state.showAll}>CP:</label>
                        <select
                            id="codingPlugSelection"
                            value={this.state.selectedCodingPlug}
                            onChange={this.handleChangeSelectedCodingPlug}
                            disabled={this.state.showAll}>

                            {Object.keys(this.cpSwToIdMapping).map(function(codingPlug, i) {
                                return <option key={i}>{codingPlug}</option>;
                            })}
                        </select>
                    </div>
                </div>
                <div className="col-3 d-flex justify-content-around">
                      <div className="form-group">
                          <label htmlFor="softwareVersionSelection" hidden={this.state.showAll}>SW:</label>
                          <select
                              id="softwareVersionSelection"
                              value={this.state.selectedSoftwareVersion}
                              onChange={this.handleChangeSelectedSoftwareVersion}
                              disabled={this.state.showAll}>

                              {Object.keys(this.cpSwToIdMapping[this.state.selectedCodingPlug]).map(function(softwareVersion, i) {
                                  return <option key={i}>{softwareVersion}</option>;
                              })}
                          </select>
                      </div>
                </div>
            </div>
        );
    }
}

export default ConfigurationHeader;
