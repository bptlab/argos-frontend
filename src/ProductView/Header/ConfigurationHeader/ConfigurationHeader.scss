.form-group {
  margin: 0 10px 5px 10px;
}
#configurationHeader select {
  background-color: inherit;
  color: $COLOR-font-light;
  border-style: none;
  width: auto;
  padding: 10px 10px;
}
#configurationHeader option {
  background-color: $COLOR-dark_background;
}
