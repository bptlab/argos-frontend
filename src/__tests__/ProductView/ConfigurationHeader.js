import React from 'react';
import ConfigurationHeader from '../../ProductView/Header/ConfigurationHeader/ConfigurationHeader.js';
import renderer from 'react-test-renderer';
import TestProduct from './../testData/product.js'

let instance, component;

const firstTestConfiguration = TestProduct.PRODUCT.configurations[0];
const secondTestConfiguration = TestProduct.PRODUCT.configurations[1];
const setProductConfiguration = jest.fn();

test("Rendering of ConfigurationHeader", () => {
    component = renderer.create(
        <ConfigurationHeader
            ref={(child) => {instance = child}}
            configurations={TestProduct.PRODUCT.configurations}
            setProductConfiguration={setProductConfiguration}/>
    );

    expect(instance.state.selectedCodingPlug).toEqual(firstTestConfiguration.codingPlugId.toString());
    expect(instance.state.selectedSoftwareVersion).toEqual(firstTestConfiguration.codingPlugSoftwareVersions[0].toString());
    expect(setProductConfiguration).toBeCalled();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Changing configuration selection", () => {
    let event = {target: {value: secondTestConfiguration.codingPlugId.toString()}};
    instance.selectedCodingPlugChanged(event);
    expect(instance.state.selectedCodingPlug).toEqual(secondTestConfiguration.codingPlugId.toString());
    expect(instance.state.selectedSoftwareVersion).toEqual(secondTestConfiguration.codingPlugSoftwareVersions[0].toString());
    expect(setProductConfiguration).toBeCalled();

    event = {target: {value: secondTestConfiguration.codingPlugSoftwareVersions[1].toString()}};
    instance.selectedSoftwareVersionChanged(event);
    expect(instance.state.selectedCodingPlug).toEqual(secondTestConfiguration.codingPlugId.toString());
    expect(instance.state.selectedSoftwareVersion).toEqual(secondTestConfiguration.codingPlugSoftwareVersions[1].toString());
    expect(setProductConfiguration).toBeCalled();
});
