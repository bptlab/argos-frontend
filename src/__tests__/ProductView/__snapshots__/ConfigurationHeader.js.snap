exports[`test Rendering of ConfigurationHeader 1`] = `
<div
  className="row col-12 d-flex justify-content-center">
  <div
    className="col-3 d-flex">
    <div
      className="form-group">
      <label
        htmlFor="codingPlugSelection">
        CP:
      </label>
      <select
        id="codingPlugSelection"
        onChange={[Function]}
        value="10">
        <option>
          10
        </option>
        <option>
          20
        </option>
        <option>
          30
        </option>
      </select>
    </div>
  </div>
  <div
    className="col-3 d-flex justify-content-around">
    <div
      className="form-group">
      <label
        htmlFor="softwareVersionSelection">
        SW:
      </label>
      <select
        id="softwareVersionSelection"
        onChange={[Function]}
        value="1">
        <option>
          1
        </option>
        <option>
          4
        </option>
        <option>
          1.1
        </option>
        <option>
          1.2
        </option>
        <option>
          4.1
        </option>
      </select>
    </div>
  </div>
</div>
`;
